### PR TITLE
EZEE-3085: Added specification ContentTypeSpecification

### PR DESCRIPTION
**Pull request by devmysz**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [EZEE-3085](https://jira.ez.no/browse/EZEE-3085)
| **Type**                                   | improvement
| **Target eZ Platform version** | `v3.0.1` - please update accordingly
| **BC breaks**                          | no
| **Tests pass**                          | yes
| **Doc needed**                       | no

Added specification ContentTypeSpecification

#### Checklist:
- [x] PR description is updated.
- [x] Tests are implemented. 
- [x] Added code follows Coding Standards (use `$ composer fix-cs`).
- [x] PR is ready for a review.

### CODE REVIEW

**Review by Nattfarinn — CHANGES_REQUESTED** [review]
Provide unit tests for `ContentTypeSpecification`.

**Review by alongosz — CHANGES_REQUESTED** [review]
Could you elaborate on the purpose of this and naming?
I don't see this doing anything right now (not configured?).

I see that naming was taken from `SpecificationInterface` defined by @mikadamczyk in AdminUI in 2018.
@mikadamczyk if you remember, could you explain the choice of naming? Is it some kind of a design pattern or does it reflect existing solution? I don't remember being part of this discussion.

Without the broader context I don't feel like this belongs to the Repository domain

**Comment by michal-myszka** [comment]
> Could you elaborate on the purpose of this and naming?
> I don't see this doing anything right now (not configured?).
> 
> I see that naming was taken from `SpecificationInterface` defined by @mikadamczyk in AdminUI in 2018.
> @mikadamczyk if you remember, could you explain the choice of naming? Is it some kind of a design pattern or does it reflect existing solution? I don't remember being part of this discussion.
> 
> Without the broader context I don't feel like this belongs to the Repository domain

This becomes from Behavioral Design Pattern - Specification:
https://designpatternsphp.readthedocs.io/en/latest/Behavioral/Specification/README.html

The `ContentTypeSpecification` is currently used in 
https://github.com/ezsystems/ezplatform-page-fieldtype/pull/141/files#diff-5adf309520410df755badb9f39017b56R51
 and should be available for the other packages.

**Comment by alongosz** [comment]
> This becomes from Behavioral Design Pattern - Specification:
> https://designpatternsphp.readthedocs.io/en/latest/Behavioral/Specification/README.html
> 
> The `ContentTypeSpecification` is currently used in
> https://github.com/ezsystems/ezplatform-page-fieldtype/pull/141/files#diff-5adf309520410df755badb9f39017b56R51
> and should be available for the other packages.

Because it's used in the following way there
```php
$contentTypeSpecification = new ContentTypeSpecification($expectedType);
```
the `ContentTypeSpecification` implementation should also be a part of SPI. We don't give BC promise on anything from Core, unless documented. Documented Core stuff is a technical debt at this point.
If you were to define `ContentTypeSpecification` in the Page Field Type package, then what we have, minus that implementation would be fine. 
We could consider `ContentTypeSpecification` an API, but then it can't be "newable" externally.

TBH it looks like an overkill, but as long as it follows A(S)PI rules, I'm fine ;)

File: eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\SPI\Specification\Tests\Content;
+
+use eZ\Publish\API\Repository\Values\Content\Content;
+use eZ\Publish\API\Repository\Values\ContentType\ContentType;
+use eZ\Publish\Core\Search\Tests\TestCase;
```
Comment (alongosz) [inline eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php:13]: Seems like an accidental import. You've meant to add the he one from PHPUnit, right? ;)

File: eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * @covers \eZ\Publish\SPI\Specification\Content\ContentContainerSpecification::isSatisfiedBy
+     * @dataProvider isSatisfiedByProvider
+     */
+    public function testIsSatisfiedBy(
+        bool $isContainer,
+        bool $shouldBeSatisfied
+    ): void
+    {
```
Comment (alongosz) [inline eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php:37]: CS
```suggestion
    ): void {
```

File: eZ/Publish/SPI/Specification/Tests/Content/ContentTypeSpecificationTest.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\SPI\Specification\Tests\Content;
+
+use eZ\Publish\API\Repository\Values\Content\Content;
+use eZ\Publish\API\Repository\Values\ContentType\ContentType;
+use eZ\Publish\Core\Search\Tests\TestCase;
```
Comment (alongosz) [inline eZ/Publish/SPI/Specification/Tests/Content/ContentTypeSpecificationTest.php:13]: Same comment as before

File: eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            ->method('getContentType')
+            ->willReturn($contentTypeMock);
+
+        $this->assertEquals(
+            $contentContainerSpecification->isSatisfiedBy($contentMock),
+            $shouldBeSatisfied
+        );
+    }
+
+    public static function isSatisfiedByProvider(): array
```
Comment (alongosz) [inline eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php:57]: Optional: when I see method starting with `is...` I immediately want to see `:bool` as a return type.
I'd avoid it, by e.g. naming it: `providerForIsSatisfiedBy` or, following verb based method naming conventions `getDataForIsSatisfiedBy`.
Both are fine for tests.
But not insisting.

File: eZ/Publish/SPI/Specification/Tests/Content/ContentTypeSpecificationTest.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            ->method('getContentType')
+            ->willReturn($contentTypeMock);
+
+        $this->assertEquals(
+            $contentTypeSpecification->isSatisfiedBy($contentMock),
+            $shouldBeSatisfied
+        );
+    }
+
+    public static function isSatisfiedByProvider(): array
```
Comment (alongosz) [inline eZ/Publish/SPI/Specification/Tests/Content/ContentTypeSpecificationTest.php:71]: Optional: the same comment

File: eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php (hunk 4 above)
Comment (alongosz) [inline eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php:57]: Question: does PHPUnit handle well `static` providers? Is it documented somewhere? Usually it's non-static.

**Review by alongosz — CHANGES_REQUESTED** [review]
Ok for SPI,
Remarks for tests:

File: eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php (hunk 1 above)
Comment (michal-myszka) [reply eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php]: Right .. :)

File: eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php (hunk 4 above)
Comment (michal-myszka) [reply eZ/Publish/SPI/Specification/Tests/Content/ContentContainerSpecificationTest.php]: I found a lot of usages in our codebase, but that's fact that in PHPUnit docs there is no any information about static call - so I've changed to non-static method.

**Review by alongosz — APPROVED** [review]

**Review by Nattfarinn — APPROVED** [review]

**Review by ViniTou — APPROVED** [review]

**Review by katarzynazawada — APPROVED** [review]

**Comment by michal-myszka** [comment]
Merged up to master

**Pull request merged**